styx-common/src/main/java/com/spotify/styx/api/Authenticator.java

```diff
@@ -0,0 +1,279 @@
+  void cacheResources() throws IOException {
+    final CloudResourceManager.Projects.List request = cloudResourceManager.projects().list();
+
+    ListProjectsResponse response;
+    do {
+      response = request.execute();
+      if (response.getProjects() == null) {
+        continue;
+      }
+      for (Project project : response.getProjects()) {
+        final boolean access = resolveProject(project);
+        logger.info("Resolved project: {}, access={}", project.getProjectId(), access);
+      }
+      request.setPageToken(response.getNextPageToken());
+    } while (response.getNextPageToken() != null);
+
+    logger.info("Resource cache loaded");
+  }
+
+  GoogleIdToken authenticate(String token) {
+    final GoogleIdToken googleIdToken;
+    try {
+      googleIdToken = verifyIdToken(token);
+    } catch (IOException e) {
+      logger.warn("Failed to verify token");
+      return null;
+    }
+
+    if (googleIdToken == null) {
+      return null;
+    }
+
+    final String email = googleIdToken.getPayload().getEmail();
+
+    final String domain = getDomain(email);
+    if (domain != null) {
+      if (domainWhitelist.contains(domain)) {
+        logger.debug("Domain {} in whitelist", domain);
+        return googleIdToken;
+      }
+    } else {
+      logger.warn("Invalid email address {}", email);
+      return null;
+    }
+
+    if (validatedEmailCache.getIfPresent(email) != null) {
+      return googleIdToken;
+    }
+
+    final String projectId;
+    try {
+      projectId = checkProject(email);
+    } catch (IOException e) {
+      logger.info("Cannot authenticate {}", email);
+      return null;
+    }
+
+    if (projectId != null) {
+      validatedEmailCache.put(email, projectId);
+      return googleIdToken;
+    }
+
+    return null;
+  }
+
+  @VisibleForTesting
+  void clearResourceCache() {
+    resourceCache.clear();
+  }
+
+  private boolean isWhitelisted(ResourceId resourceId) {
+    return resourceWhitelist.contains(resourceId) || resourceCache.contains(resourceId);
+  }
+
+  @VisibleForTesting
+  static ResourceId resourceId(String type, String id) {
+    return new ResourceId().setType(type).setId(id);
+  }
+
+  @VisibleForTesting
+  static ResourceId resourceId(Project project) {
+    return resourceId("project", project.getProjectId());
+  }
+
+  private GoogleIdToken verifyIdToken(String token) throws IOException {
+    try {
+      return googleIdTokenVerifier.verify(token);
+    } catch (GeneralSecurityException e) {
+      return null;
+    }
+  }
+
+  private String checkProject(String email) throws IOException {
+    String projectId = null;
+
+    final Matcher matcher = SERVICE_ACCOUNT_PATTERN.matcher(email);
+    if (matcher.matches()) {
+      projectId = matcher.group(1);
+    }
+
+    if (projectId == null) {
+      // no projectId, could be GCE default
+      logger.debug("Email {} doesn't contain project id, looking up its project", email);
+      projectId = getProjectIdOfServiceAccount(email);
```

Comment:
This could also be some random email address `foo@bar.com`, and now we end up calling `serviceAccounts.get` for those email addresses. 😕

Reply:
https://github.com/spotify/styx/pull/619